scripts/x-api.js

```diff
@@ -0,0 +1,596 @@
+      if (typeof playbackSize !== "undefined") {
+        extensions["https://w3id.org/xapi/video/extensions/video-playback-size"] = playbackSize
+      }
+
+      return {
+        "verb": {
+          "id": "http://adlnet.gov/expapi/verbs/interacted",
+          "display": {
+            "en-US": "interacted"
+          }
+        },
+        "object": getXAPIObject(),
+        "result": {
+          "extensions": {
+            "https://w3id.org/xapi/video/extensions/time": resultExtTime
+          }
+        },
+        "context": {
+          "contextActivities": {
+            "category": [{
+              "id": "https://w3id.org/xapi/video"
+            }]
+          },
+          "extensions": extensions
+        },
+        "timestamp" : timeStamp
+      };
+    };
+
+    /**
+     * Generates "completed" xAPI statement (Video Profile).
+     * @see https://liveaspankaj.gitbooks.io/xapi-video-profile/content/statement_data_model.html#236-completed
+     *
+     * @public
+     * @param {Number} currentTime time of the video currently
+     * @param {Number} duration length of the current video in seconds
+     * @returns {Object} JSON xAPI statement
+     */
+    self.getArgsXAPICompleted = function (currentTime, duration) {
+      var progress = self.getProgress(currentTime, duration);
+      var resultExtTime = formatFloat(currentTime);
+      var dateTime = new Date();
+      endPlayedSegment(resultExtTime);
+      var timeStamp = dateTime.toISOString();
+
+      var extensions = {};
+      if (typeof resultExtTime !== "undefined") {
+        extensions["https://w3id.org/xapi/video/extensions/time"] = resultExtTime
+      }
+      if (typeof progress !== "undefined") {
+        extensions["https://w3id.org/xapi/video/extensions/progress"] = progress
+      }
+      if (typeof playedSegments !== "undefined") {
+        extensions["https://w3id.org/xapi/video/extensions/played-segments"] = playedSegments
+      }
+
+      return {
+        "verb": {
+          "id": "http://adlnet.gov/expapi/verbs/completed",
+          "display": {
+            "en-US": "completed"
+          }
+        },
+        "object": getXAPIObject(),
+        "result": {
+          "extensions": extensions,
+          "success" : true,
+          "duration" : secondsToISO8601Duration(duration)
+        },
+        "context": {
+          "contextActivities": {
+            "category": [{
+              "id": "https://w3id.org/xapi/video"
+            }]
+          },
+          "extensions": {
+            "https://w3id.org/xapi/video/extensions/session-id": sessionID
+          }
+        },
+        "timestamp" : timeStamp
+      };
+    };
+
+    /**
+     * Calculate video progress.
+     *
+     * @public
+     * @param {Number} currentTime current time of the video in seconds
+     * @param {Number} duration legnth of the video in seconds
+     * @returns {Number} Progress between 0..1
+     */
+    self.getProgress = function (currentTime, duration) {
```

Comment:
Let's pass this suggestion by @jhaag75 first, since it is from his reference implementation:
https://github.com/jhaag75/xapi-videojs/blob/master/xapi-videojs.js#L51-L91

Reply:
Refactored to use `map()` and `reduce()` here: https://github.com/h5p/h5p-video/pull/17/commits/a00b8b5a25a5f38f372508836d413ff8d5e765c5